File: src/game/systems/respawn.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Checkpoint = require 'game.components.checkpoint'
local Fixture = require 'game.components.fixture'
local Player = require 'game.components.player'
local Position = require 'game.components.position'
local Respawn = require 'game.components.respawn'
local Timer = require 'game.components.timer'

local aspect = Aspect.new({Respawn, Timer})
local RespawnSystem = System:new('respawn', aspect)

function findCheckpoint(entity)
  local player = entity:as(Player)
  local fixture = entity:as(Fixture)

  for _, entity in pairs(entity.manager.entities) do
    local checkpoint = entity:as(Checkpoint)

    if checkpoint and player.checkpoint == checkpoint.index then
      return entity
    end
  end
end

function RespawnSystem:update(dt)
  for _, entity in pairs(self.entities) do
    local player = entity:as(Player)
    local respawn = entity:as(Respawn)
    local timers = entity:as(Timer).timers

    if timers.respawn == nil then
      player.lives = math.max(0, player.lives - 1)
      timers.respawn = {waitTime = respawn.waitTime}
    elseif timers.respawn and timers.respawn.waitTime == 0 then
      timers.respawn = nil

      local checkpoint = findCheckpoint(entity)

      if checkpoint then
        local fixture = entity:as(Fixture)
        local body = fixture.fixture:getBody()
        local position = checkpoint:as(Position)

        body:setType('static')
        body:setPosition(position.x, position.y)
        body:setType('dynamic')
      end

      entity:remove(Respawn)
    end
  end
end

return RespawnSystem

File: client2/src/game/systems/Respawn.ts
import { Aspect } from '../../ecs/Aspect'
import { Entity } from '../../ecs/Entity'
import { Manager } from '../../ecs/manager'
import { System } from '../../ecs/System'
import { Checkpoint } from '../components/Checkpoint'
import { GameObject } from '../components/GameObject'
import { Health } from '../components/Health'
import { Player } from '../components/Player'
import { Position } from '../components/Position'
import { Respawn } from '../components/Respawn'
import { SystemFlag } from '../flags'
import { setTimeout } from '../utils/timer'

const findCheckpoint = (
  index: number,
  manager: Manager
): Entity | undefined => {
  let entity

  manager.entities.forEach(entity => {
    const checkpoint = entity.as(Checkpoint)

    if (checkpoint && checkpoint.index === index) {
      entity = entity
    }
  })

  return entity
}

const resetPosition = (entity: Entity) => {
  const player = entity.as(Player)
  const health = entity.as(Health)

  if (!player || !health) {
    return
  }

  const checkpoint = findCheckpoint(
    player.checkpoint,
    entity.manager as Manager
  )

  if (!checkpoint) {
    return
  }

  const gameObject = entity.as(GameObject)
  const position = checkpoint.as(Position)

  if (!position || !gameObject) {
    return
  }

  const body = gameObject.fixture.getBody()
  body.setType('static')
  body.setPosition(position.x, position.y)
  body.setType('dynamic')
}

export class RespawnSystem extends System {
  static _id = 'Respawn'
  _id = RespawnSystem._id

  static _flag = SystemFlag.Respawn
  _flag = RespawnSystem._flag

  static _aspect = new Aspect([Respawn])
  _aspect = RespawnSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const player = entity.as(Player)
      const respawn = entity.as(Respawn)
      const health = entity.as(Health)

      if (!player || !respawn || !health) {
        return
      }

      if (health.hitpoints === 0) {
        respawn.timer = setTimeout(respawn.duration, () => {
          resetPosition(entity)
          entity.remove(Respawn)
        })
      }

      health.hitpoints = 1
    })
  }
}
